### PR TITLE
Add securityContext option to agent

### PR DESCRIPTION
**Pull request by tenshinhigashi**
Signed-off-by: tenshinhigashi <thigashi@datawire.io>

## Description

Adding securityContexts to the Agent and apiext. The agent security contexts can be edited in the Helm files with `.Values.agent.containerSecurityContext` and `.Values.agent.podSecurityContext`. 

## Testing

Most of the testing has happened locally. Using `make deploy` I have confirmed that the expected deployments are updated successfully. 

I have tested agent by connecting to the Cloud as expected.
I have tested the emissary-apiext by applying a `v3alpha1` resource and confirming that it still translates as expected.

## Checklist

- [ ] **Does my change need to be backported to a previous release?**
  - What backport versions were discussed with the Maintainers in the Issue?

- [x] **I made sure to update `CHANGELOG.md`.**

   Remember, the CHANGELOG needs to mention:
  - Any new features
  - Any changes to our included version of Envoy
  - Any non-backward-compatible changes
  - Any deprecations

- [x] **This is unlikely to impact how Ambassador performs at scale.**

   Remember, things that might have an impact at scale include:
  - Any significant changes in memory use that might require adjusting the memory limits
  - Any significant changes in CPU use that might require adjusting the CPU limits
  - Anything that might change how many replicas users should use
  - Changes that impact data-plane latency/scalability

- [x] **My change is adequately tested.**

   Remember when considering testing:
  - Your change needs to be specifically covered by tests.
    - Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
  - You also need to make sure that the _entire area being changed_ has adequate test coverage.
    - If existing tests don't actually cover the entire area being changed, add tests.
    - This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
  - We should lean on the bulk of code being covered by unit tests, but...
  - ... an end-to-end test should cover the integration points

- [ ] **I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.**

- [ ] **The changes in this PR have been reviewed for security concerns and adherence to security best practices.**

### CODE REVIEW

**Comment by LanceEa** [comment]
@tenshinhigashi - since this is just a WIP for testing in CI, you can leverage "draft" PR's capability instead of using WIP in title.  https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Review by LanceEa — COMMENTED** [review]
I have just a few general questions that I want to make sure we consider. Also, did we pull this into AES and verify it doesn't break Edge Stack?

File: charts/emissary-ingress/values.yaml.in [hunk 1]
```diff
@@ -404,6 +404,12 @@ agent:
     tag: 3.1.0
     repository: docker.io/emissaryingress/emissary
 
+  podSecurityContext:
```
[the remark this answers is not in the document]
Comment (tenshinhigashi) [reply charts/emissary-ingress/values.yaml.in:407]: From my quick testing:
`podSecurityContext` will default to `{}`. 
`containerSecurityContext` will simply not show up.

I have also quickly tested the same configuration with the standalone agent and everything worked as expected. 

As far as I can tell, this PR should be good to go?

**Comment by LanceEa** [comment]
@tenshinhigashi - LGTM, I think we just need to make sure AES Redis plays nice with PSP too now.

Also, what version of K8s did you run your tests against? Did you test against a 1.25 cluster since PSP is removed there? My guess is your change here works perfectly fine but that if a user was to enable `security.podSecurityPolicy` in the helm chart the helm install would probably fail due to not being able to apply  the `PodSecurityPolicy`.

**Review by LanceEa — DISMISSED** [review]

**Comment by joebowbeer** [comment]
Add tests for this? For example, see cert-manager e2e tests with an admission controller. (Though these days you don't even need to install an admission controller; you can just enable PSA in your namespace.)

You can verify this in static checks using kyverno CLI.

**Comment by LanceEa** [comment]
@tenshinhigashi - It looks like your rebase pulled in some additional commits. I can help you fix this when you are back from vacation.

@joebowbeer - Always happy to add additional test coverage, would you be open to submitting a PR once this lands?

**Comment by joebowbeer** [comment]
To meet restricted-level pod security standards on os=linux, each container needs the equivalent of the following, some of which can be declared in the pod-level security context:

```yaml
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          readOnlyRootFilesystem: true
          runAsNonRoot: true
          seccompProfile:
            type: RuntimeDefault
```

Why are only two of these included (commented-out) in values?

Why not make these the defaults?

**Comment by LanceEa** [comment]
>Why not make these the defaults?

@joebowbeer - To ensure that we don't break anyone during an upgrade, we have chosen to approach this as an opt-in by providing the knobs in helm for folks that want to enable these additional settings. The comments are just meant to be examples and not an exhaustive list of what K8s supports in those fields. 

@tenshinhigashi - one last nit. I think we can make the commented out lines clearer that the values accepted in these settings are based on accepted values in K8s: For example, we provide a link to give users indication as seen here so we should be able to update this: https://github.com/emissary-ingress/emissary/blob/edfd64ac747b8176c7b91dbce2d5372bd466ea3c/charts/emissary-ingress/values.yaml.in#L108

**Comment by joebowbeer** [comment]
@LanceEa Thanks for clarifying. If there are any recommended restricted-mode settings (see my list) that will _not_ work, please call them out.

Is it worth adding a deprecation comment for PSP?

> PodSecurityPolicy was [deprecated](https://kubernetes.io/blog/2021/04/08/kubernetes-1-21-release-announcement/#podsecuritypolicy-deprecation) in Kubernetes v1.21, and removed from Kubernetes in v1.25.

**Review by LanceEa — APPROVED** [review]
lgtm...lets squash merge the commits and cleanup the messages. Let me know if you have questions on that.

**Comment by LanceEa** [comment]
> @LanceEa Thanks for clarifying. If there are any recommended restricted-mode settings (see my list) that will _not_ work, please call them out.
> 
> Is it worth adding a deprecation comment for PSP?
> 
> > PodSecurityPolicy was [deprecated](https://kubernetes.io/blog/2021/04/08/kubernetes-1-21-release-announcement/#podsecuritypolicy-deprecation) in Kubernetes v1.21, and removed from Kubernetes in v1.25.

@joebowbeer There are no issues that I'm aware of but I'm not a PSP expert so happy to hear other voices on the subject. 

As for k8s 1.25 deprecation notice, I'm happy to accept a PR for this if you have ideas on where it can be called out.

**Comment by AliceProxy** [comment]
@LanceEa 
> As for k8s 1.25 deprecation notice, I'm happy to accept a PR for this if you have ideas on where it can be called out.

This can just be a comment above the helm value warning users that it is deprecated in 1.25, beyond that the field is opt-in so its up to the user to determine whether they should use it or not. We have plenty of users running older versions of Kubernetes that will not upgrade in any immediate timeline.

**Review by AliceProxy — APPROVED** [review]
lgtm

**Pull request merged**